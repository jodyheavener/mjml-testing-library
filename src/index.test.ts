import { JSDOM } from 'jsdom';
import mjml2html from 'mjml';
import { MJMLJsonWithChildren } from 'mjml-core';
import { cleanup, render, screen } from './';

const exampleText = 'This is some text';

const unwrappedTemplate = `
  <mj-section>
    <mj-column>
      <mj-text color="blue" font-size="10px">
        ${exampleText}
      </mj-text>
    </mj-column>
  </mj-section>
`;

const wrappedTemplate = `
  <mjml>
    <mj-body>
      ${unwrappedTemplate}
    </mj-body>
  </mjml>
`;

describe('render', () => {
  afterEach(cleanup);

  it('renders the template into the document', () => {
    const renderedHtml = mjml2html(wrappedTemplate).html;
    const body = new JSDOM(renderedHtml).window.document.body;

    const { container } = render(wrappedTemplate);
    expect(container.innerHTML).toEqual(body.innerHTML);
    expect(container).toBeInTheDocument();
  });

  it('applies MJML configuration', () => {
    // Let's set the validationLevel to strict and then try to
    // parse some invalid MJML; the error should be expected
    expect(() => {
      render(
        `
          <mjml>
            <mj-column>
              <mj-section>
                Invalid syntax
              </mj-section>
            </mj-column>
          </mjml>
        `,
        {
          mjmlOptions: { validationLevel: 'strict' },
        }
      );
    }).toThrow();
  });

  it('returns the template json', () => {
    const renderedJson = mjml2html(wrappedTemplate).json;

    const { json } = render(wrappedTemplate);
    expect(json).toStrictEqual(renderedJson);
    expect(json).toMatchSnapshot();
  });

  it('can wrap the template with base mjml components', () => {
    const { json, container } = render(unwrappedTemplate, {
      wrap: true,
      mjmlOptions: { validationLevel: 'strict' },
    });

    const body = (json as MJMLJsonWithChildren).children[0];
    expect(body.tagName).toEqual('mj-body');
    expect(container).toMatchSnapshot();
  });

  it('returns baseElement which defaults to document.body', () => {
    const { baseElement } = render(wrappedTemplate);
    expect(baseElement).toBe(document.body);
  });

  it('supports fragments', () => {
    const { asFragment } = render(wrappedTemplate);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('debug', () => {
    beforeEach(() => {
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      jest.spyOn(console, 'log').mockImplementation(() => {});
    });

    afterEach(() => {
      (console.log as jest.Mock).mockRestore();
      cleanup();
    });

    it('pretty prints the container', () => {
      const { debug } = render(wrappedTemplate);

      debug();

      expect(console.log).toHaveBeenCalledTimes(1);
      expect(console.log).toHaveBeenCalledWith(
        expect.stringContaining(exampleText)
      );
    });

    it('pretty prints multiple container', () => {
      const { debug } = render(
        `
          <mjml>
            <mj-body>
              <mj-section>
                <mj-column>
                  <mj-text color="blue" font-size="10px">
                    ${exampleText}
                  </mj-text>
                  <mj-text color="blue" font-size="10px">
                    ${exampleText}
                  </mj-text>
                </mj-column>
              </mj-section>
            </mj-body>
          </mjml>
        `
      );

      const multipleElements = screen.getAllByText(exampleText);
      debug(multipleElements);

      expect(console.log).toHaveBeenCalledTimes(2);
      expect(console.log).toHaveBeenCalledWith(
        expect.stringContaining(exampleText)
      );
    });
  });
});

describe('cleanup', () => {
  it('cleans up the document', () => {
    render(wrappedTemplate);
    expect(screen.queryByText(exampleText)).toBeInTheDocument();

    cleanup();

    expect(screen.queryByText(exampleText)).not.toBeInTheDocument();
    expect(document.body).toBeEmptyDOMElement();
  });

  it('does not error when an element is not a child', () => {
    render(wrappedTemplate, { container: document.createElement('div') });
    cleanup();
  });
});
