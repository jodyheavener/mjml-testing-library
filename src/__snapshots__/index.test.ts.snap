// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render can wrap the template with base mjml components 1`] = `
<div>
  
        
        
      
  <div
    style=""
  >
    
        
      
      
    <!--[if mso | IE]&gt;&lt;table align="center" border="0" cellpadding="0" cellspacing="0" class="" role="presentation" style="width:600px;" width="600" &gt;&lt;tr&gt;&lt;td style="line-height:0px;font-size:0px;mso-line-height-rule:exactly;"&gt;&lt;![endif]-->
    
    
      
      
    <div
      style="margin:0px auto;max-width:600px;"
    >
      
        
        
      <table
        align="center"
        border="0"
        cellpadding="0"
        cellspacing="0"
        role="presentation"
        style="width:100%;"
      >
        
          
        <tbody>
          
            
          <tr>
            
              
            <td
              style="direction:ltr;font-size:0px;padding:20px 0;text-align:center;"
            >
              
                
              <!--[if mso | IE]&gt;&lt;table role="presentation" border="0" cellpadding="0" cellspacing="0"&gt;&lt;tr&gt;&lt;td class="" style="vertical-align:top;width:600px;" &gt;&lt;![endif]-->
              
            
      
              <div
                class="mj-column-per-100 mj-outlook-group-fix"
                style="font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;"
              >
                
        
      
                <table
                  border="0"
                  cellpadding="0"
                  cellspacing="0"
                  role="presentation"
                  style="vertical-align:top;"
                  width="100%"
                >
                  
        
                  <tbody>
                    
          
              
                    <tr>
                      
                
                      <td
                        align="left"
                        style="font-size:0px;padding:10px 25px;word-break:break-word;"
                      >
                        
                  
      
                        <div
                          style="font-family:Ubuntu, Helvetica, Arial, sans-serif;font-size:10px;line-height:1;text-align:left;color:blue;"
                        >
                          This is some text
                        </div>
                        
    
                
                      </td>
                      
              
                    </tr>
                    
            
        
                  </tbody>
                  
      
                </table>
                
    
      
              </div>
              
    
          
              <!--[if mso | IE]&gt;&lt;/td&gt;&lt;/tr&gt;&lt;/table&gt;&lt;![endif]-->
              
              
            </td>
            
            
          </tr>
          
          
        </tbody>
        
        
      </table>
      
        
      
    </div>
    
    
      
      
    <!--[if mso | IE]&gt;&lt;/td&gt;&lt;/tr&gt;&lt;/table&gt;&lt;![endif]-->
    
    
    
      
  </div>
  
    
      
    
  
</div>
`;

exports[`render returns the template json 1`] = `
Object {
  "absoluteFilePath": "/Users/jody/Downloads/Repositories/mjml-testing-library",
  "attributes": Object {},
  "children": Array [
    Object {
      "absoluteFilePath": "/Users/jody/Downloads/Repositories/mjml-testing-library",
      "attributes": Object {},
      "children": Array [
        Object {
          "absoluteFilePath": "/Users/jody/Downloads/Repositories/mjml-testing-library",
          "attributes": Object {},
          "children": Array [
            Object {
              "absoluteFilePath": "/Users/jody/Downloads/Repositories/mjml-testing-library",
              "attributes": Object {},
              "children": Array [
                Object {
                  "absoluteFilePath": "/Users/jody/Downloads/Repositories/mjml-testing-library",
                  "attributes": Object {
                    "color": "blue",
                    "font-size": "10px",
                  },
                  "content": "This is some text",
                  "file": ".",
                  "includedIn": Array [],
                  "line": 7,
                  "tagName": "mj-text",
                },
              ],
              "file": ".",
              "includedIn": Array [],
              "line": 6,
              "tagName": "mj-column",
            },
          ],
          "file": ".",
          "includedIn": Array [],
          "line": 5,
          "tagName": "mj-section",
        },
      ],
      "file": ".",
      "includedIn": Array [],
      "line": 3,
      "tagName": "mj-body",
    },
  ],
  "file": ".",
  "includedIn": Array [],
  "line": 2,
  "tagName": "mjml",
}
`;

exports[`render supports fragments 1`] = `
<DocumentFragment>
  
        
        
      
  <div
    style=""
  >
    
        
      
      
    <!--[if mso | IE]&gt;&lt;table align="center" border="0" cellpadding="0" cellspacing="0" class="" role="presentation" style="width:600px;" width="600" &gt;&lt;tr&gt;&lt;td style="line-height:0px;font-size:0px;mso-line-height-rule:exactly;"&gt;&lt;![endif]-->
    
    
      
      
    <div
      style="margin:0px auto;max-width:600px;"
    >
      
        
        
      <table
        align="center"
        border="0"
        cellpadding="0"
        cellspacing="0"
        role="presentation"
        style="width:100%;"
      >
        
          
        <tbody>
          
            
          <tr>
            
              
            <td
              style="direction:ltr;font-size:0px;padding:20px 0;text-align:center;"
            >
              
                
              <!--[if mso | IE]&gt;&lt;table role="presentation" border="0" cellpadding="0" cellspacing="0"&gt;&lt;tr&gt;&lt;td class="" style="vertical-align:top;width:600px;" &gt;&lt;![endif]-->
              
            
      
              <div
                class="mj-column-per-100 mj-outlook-group-fix"
                style="font-size:0px;text-align:left;direction:ltr;display:inline-block;vertical-align:top;width:100%;"
              >
                
        
      
                <table
                  border="0"
                  cellpadding="0"
                  cellspacing="0"
                  role="presentation"
                  style="vertical-align:top;"
                  width="100%"
                >
                  
        
                  <tbody>
                    
          
              
                    <tr>
                      
                
                      <td
                        align="left"
                        style="font-size:0px;padding:10px 25px;word-break:break-word;"
                      >
                        
                  
      
                        <div
                          style="font-family:Ubuntu, Helvetica, Arial, sans-serif;font-size:10px;line-height:1;text-align:left;color:blue;"
                        >
                          This is some text
                        </div>
                        
    
                
                      </td>
                      
              
                    </tr>
                    
            
        
                  </tbody>
                  
      
                </table>
                
    
      
              </div>
              
    
          
              <!--[if mso | IE]&gt;&lt;/td&gt;&lt;/tr&gt;&lt;/table&gt;&lt;![endif]-->
              
              
            </td>
            
            
          </tr>
          
          
        </tbody>
        
        
      </table>
      
        
      
    </div>
    
    
      
      
    <!--[if mso | IE]&gt;&lt;/td&gt;&lt;/tr&gt;&lt;/table&gt;&lt;![endif]-->
    
    
    
      
  </div>
  
    
      
    
  
</DocumentFragment>
`;
